### ISSUE TITLE
Location of too-many-public-methods error outside of existing code 

### ISSUE THREAD

**Issue by mbdevpl** [PyCQA/pylint#1839]
### Steps to reproduce
1. Prepare an "adversarial file" with too many public methods, [e.g. this one](https://gist.github.com/mbdevpl/5e33cb1c23c5d8e011f8e656eb0883a8#file-pylint_wrong_error_loc-py) - which is a cut down version of [this one](https://github.com/mbdevpl/version-query/blob/master/version_query/version.py).

   I call it adversarial, because so far I've encountered this error with only one file (and its derivatives), and the occurrence of error depends on something really obscure. 

2. Run pylint on the file.

### Current behavior

Among pylint's output there's a line referring to a location that is not existing in the code:
```
R: 34,31: Too many public methods (25/20) (too-many-public-methods)
```
([full log here](https://gist.github.com/mbdevpl/5e33cb1c23c5d8e011f8e656eb0883a8#file-wrong_loc-log-L22)), but line 34 of checked file has only 14 characters ([see here](https://gist.github.com/mbdevpl/5e33cb1c23c5d8e011f8e656eb0883a8#file-pylint_wrong_error_loc_fixed-py-L34)) which causes errors in tools that depend on pylint reporting correct error locations, e.g. AtomLinter/linter-pylint#235.

### Expected behavior

What fixes the issue for the case above, is commenting out a seemingly unrelated function `__repr__` in the checked file ([as seen here](https://gist.github.com/mbdevpl/5e33cb1c23c5d8e011f8e656eb0883a8#file-pylint_wrong_error_loc_fixed-py-L216))

Then the error becomes, as expected:
```
R: 34,0: Too many public methods (25/20) (too-many-public-methods)
```
([full log here](https://gist.github.com/mbdevpl/5e33cb1c23c5d8e011f8e656eb0883a8#file-wrong_loc_fixed-log-L22)).

### pylint --version output

```
No config file found, using default configuration
pylint 1.8.1, 
astroid 1.6.0
Python 3.6.3 (default, Oct 18 2017, 23:31:34) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.38)]
```

**Comment by brycepg** [comment]
Strange.. It has something to do with the __repr__

**Comment by brycepg** [comment]
I can reproduce this error with the following when setting max_classes=-1
```python
class F:
    def _(self):
        return type(self).f
```

**Comment by brycepg** [comment]
So for some reason inference of the attribute `f` (in my example) in `TypeChecker.visit_attribute` (`checkers/typecheck.py:696`)  causes the `col_offset` offset of the class be set to the imagined attribute `col_offset` of `27`. I don't expect to fix this soon because my understanding of the inference system is limited and I don't know how to debug it well.

**Comment by PCManticore** [comment]
I don't remember exactly why we set the col_offset of the result. But in your example, @brycepg, I bet this happens just for `type()`. If you check how we infer `type()` calls, we basically call `object_type` which just returns in case of a class, its metaclass. I presume that is going to be a singleton object, so the code you linked from `brain_builtin_inference` would just modify the `col_offset` of the metaclass. Didn't test to see if this is the case, but worth taking a look.

https://github.com/PyCQA/astroid/blob/master/astroid/helpers.py#L52

**Comment by brycepg** [comment]
I don't quite understand the inference path, but I'm only seeing an `Instance` of F object in `_object_type` (when inferring the self `Name`). I think this is because the `_explicit_inference` of the type `Call` is set to the transform wrapper (what I linked) in the inference_tip. Changing the line I linked fixes this issue (I've also found this issue affecting `getattr` calls)

**Comment by PCManticore** [comment]
@brycepg I don't understand your question regarding the node existing in the ast, can you clarify a bit?
Regarding your last statement, that makes sense!
